### ISSUE TITLE
node restart (0 peers, does not sync, out of sync)

### ISSUE THREAD

**Issue by antiochp** [mimblewimble/grin#294]
Looks like we have an issue when restarting a node.
It is possible for the node to see no peers, causing the sync to end without actually syncing.
Node never catches up to the network.
And if the node is mining it starts mining a fork.

[will add logs when I reproduce this again]

Related - #283.

**Comment by antiochp** [comment]
@ignopeverell 

From debugging locally I can see `connect_and_req` called (sometimes but not consistently) at startup but then it doesn't appear to get called again, even when it is monitoring for new peers and we are still below the desired threshold of peers.

It feels like `listen_for_addrs` loops for while but then stops or exits.

I'm not familiar enough with the UnboundedReceiver to know if this is the case though.

**Comment by antiochp** [comment]
Said another way - we're seeing logs that suggest addrs are being added to the UnboundedSender, but we are not seeing logs that show anything is being read on the UnboundedReceiver after a period of time, it just seems to stop.

**Comment by antiochp** [comment]
Clearing out `.grin/peers` appears to resolve this - but I don't fully understand how or why.

**Comment by antiochp** [comment]
If we get into a state where this is logged during the 10s sync window then we fail to sync with the network - 
```
monitoring peers (0)
```

But I don't understand why this would _ever_ happen when we see approx 80 peers in the peer_store db.

**Comment by ignopeverell** [comment]
Do you have an easy reproduction scenario? Just start a node, sync, stop and restart?

**Comment by antiochp** [comment]
Yeah that's pretty much it.

1) The `monitoring peers (n)` log msg shows how many peers we are currently connected to.
1) We monitor for peers every 10s
1) If we have less than `PEER_PREFERRED_COUNT` then we should add more
1) but we don't appear to add more

I would expect the log to say -
`monitoring peers (8)` (after a few monitoring cycles after startup)
But we often seem to get stuck with `n<8`.
And if it is `n==0` then we get into trouble (but that is not easily reproducible).

**Comment by antiochp** [comment]
Just restarted a node and got this - 
```
Nov 17 14:05:31.650 DEBG monitoring peers (3)
Nov 17 14:05:31.651 DEBG Got 47 more peers from db, trying to connect.
Nov 17 14:05:41.650 DEBG monitoring peers (3)
Nov 17 14:05:41.651 DEBG Got 47 more peers from db, trying to connect.
Nov 17 14:05:51.650 DEBG monitoring peers (3)
Nov 17 14:05:51.651 DEBG Got 47 more peers from db, trying to connect.
```

If we have 47 peers in the db, why are we stuck at 3 connected peers?

**Comment by ignopeverell** [comment]
Knowing why would likely solve this issue :-) My hunch would be a clogged/stuck channel or channel listener but I'm trying to setup a listening-only server first to be able to reproduce these.

**Comment by antiochp** [comment]
I've got debug statements and println! all over my local node and may have just found something - 

I see it entering `connect_and_req`
but I don't see it finishing.
And if we wait long enough we get this - 
```
Nov 17 14:17:19.463 ERRO Peer request error: Connection(Error { repr: Os { code: 60, message: "Operation timed out" } })
```

And while this was happening the rx/tx channel is filling up with more and more peer addrs to try.

**Comment by antiochp** [comment]
Resolved in #296. Up for discussion. @ignopeverell

**Comment by dongcarl** [comment]
Should close?
